apps/rush-lib/src/utilities/Utilities.ts

```diff
@@ -442,6 +453,59 @@ export class Utilities {
       });
   }
 
+  /**
+   * NPM allows environment variables to be specified in its .npmrc file.  This can be
+   * used to provide an authentication token for a registry.  However if the environment variable
+   * is undefined, it expands to an empty string, which produces a valid-looking mapping
+   * with an invalid URL.  As a workaround, _syncNpmrc() copies the .npmrc file to the
+   * target folder  and strips out any lines that reference undefined environment variables.
+   * If the target folder does not exist, then _syncNpmrc() will delete an .npmrc that
+   * is found there.
+   *
+   * IMPORTANT: THIS CODE SHOULD BE KEPT UP TO DATE WITH _syncNpmrc() FROM scripts/install-run.ts
+   */
+  private static _syncNpmrc(sourceNpmrcFolder: string, targetNpmrcFolder: string): void {
+    const sourceNpmrcPath: string = path.join(sourceNpmrcFolder, '.npmrc');
+    const targetNpmrcPath: string = path.join(targetNpmrcFolder, '.npmrc');
+    try {
+      if (fsx.existsSync(sourceNpmrcPath)) {
+        console.log(`Copying ${sourceNpmrcPath} --> ${targetNpmrcPath}`);
+        let npmrcFileLines: string[] = fsx.readFileSync(sourceNpmrcPath).toString().split('\n');
+        npmrcFileLines = npmrcFileLines.map((line) => (line || '').trim());
+        const resultLines: string[] = [];
+        // Trim out lines that reference environment variables that aren't defined
+        for (const line of npmrcFileLines) {
+          // This finds environment variable tokens that look like "${VAR_NAME}"
+          const regex: RegExp = /\$\{([^\}]+)\}/g;
+          const environmentVariables: string[] | null = line.match(regex);
+          let lineShouldBeTrimmed: boolean = false;
+          if (environmentVariables) {
+            for (const token of environmentVariables) {
+              // Remove the leading "${" and the trailing "}" from the token
+              const environmentVariableName: string = token.substring(2, token.length - 1);
+              if (!process.env[environmentVariableName]) {
```

Comment:
>process.env [](start = 19, length = 11)

Do you want to pass in the env variables?

Reply:
Eventually, but I think we should finalize the environment variable scrubbing feature before it's worth it

---
In reply to: [195888131](https://github.com/Microsoft/web-build-tools/pull/700#discussion_r195888131) [](ancestors = 195888131)